robots/bebop.h

```diff
@@ -0,0 +1,1188 @@
+        Limits<UnitType> m_Limits;
+        Semaphore m_Semaphore;
+    };
+
+    ControllerPtr m_Device;
+    Semaphore m_Semaphore;
+    std::unique_ptr<VideoStream> m_VideoStream;
+    bool m_IsConnected = false;
+    FlightEventHandler m_FlightEventHandler = nullptr;
+    LimitValues<degree_t> m_TiltLimits;
+    LimitValues<meters_per_second_t> m_VerticalSpeedLimits;
+    LimitValues<degrees_per_second_t> m_YawSpeedLimits;
+
+    void startStreaming();
+    void stopStreaming();
+    bool onAxisEvent(const HID::JAxis axis, const float value);
+    bool onButtonEvent(const HID::JButton button, const bool pressed);
+    inline void addEventHandlers();
+    void onBatteryChanged(const ARCONTROLLER_DICTIONARY_ELEMENT_t *dict);
+    inline void createControllerDevice();
+    inline eARCONTROLLER_DEVICE_STATE getState();
+    inline eARCONTROLLER_DEVICE_STATE getStateUpdate();
+
+    // speed limits
+    inline void setMaximumTilt(const degree_t newValue);
+    inline void setMaximumVerticalSpeed(const meters_per_second_t newValue);
+    inline void setMaximumYawSpeed(const degrees_per_second_t newValue);
+
+    static void commandReceived(eARCONTROLLER_DICTIONARY_KEY key,
+                                ARCONTROLLER_DICTIONARY_ELEMENT_t *dict,
+                                void *data);
+    static int printCallback(eARSAL_PRINT_LEVEL level,
+                             const char *tag,
+                             const char *format,
+                             va_list va);
+    static void stateChanged(eARCONTROLLER_DEVICE_STATE newstate,
+                             eARCONTROLLER_ERROR err,
+                             void *data);
+}; // Bebop
+
+// We also have to give definitions for these variables, because c++ is weird
+constexpr degree_t Bebop::DefaultMaximumTilt;
+constexpr degrees_per_second_t Bebop::DefaultMaximumYawSpeed;
+constexpr meters_per_second_t Bebop::DefaultMaximumVerticalSpeed;
+
+/*
+ * Do all initialisation (including discovery) but don't actually
+ * connect to drone yet.
+ */
+Bebop::Bebop(const degrees_per_second_t maxYawSpeed,
+             const meters_per_second_t maxVerticalSpeed,
+             const degree_t maxTilt)
+{
+    // silence annoying messages printed by library
+    ARSAL_Print_SetCallback(printCallback);
+
+    // object to interface with drone
+    createControllerDevice();
+
+    // to handle changes in state, incoming commands
+    addEventHandlers();
+
+    // initialise video stream object
+    m_VideoStream = std::make_unique<VideoStream>(*this);
+
+    // store speed limits for later
+    m_YawSpeedLimits.m_UserMaximum = maxYawSpeed;
+    m_VerticalSpeedLimits.m_UserMaximum = maxVerticalSpeed;
+    m_TiltLimits.m_UserMaximum = maxTilt;
+}
+
+/*
+ * Disconnect properly from drone when object destroyed.
+ */
+Bebop::~Bebop()
+{
+    land();
+    stopStreaming();
+    disconnect();
+}
+
+/*
+ * Try to make connection to drone.
+ */
+void
+Bebop::connect()
+{
+    if (m_IsConnected) {
```

Comment:
If you connected in the constructor (as there only seems to be one) you could skip this logic too :)